akka-actor/src/main/scala/akka/actor/CoordinatedShutdown.scala

```diff
@@ -354,16 +351,120 @@ final class CoordinatedShutdown private[akka] (
+        // This is a vanilla class instead of a case class to avoid default implementations of .hashCode() and .equals
+        // Different code paths could register the same task under the same name multiple times; in this case we want to
+        // run that task as many times as it was registered (minus the number of times those were cancelled), so they must
+        // be distinct in a Set[TaskDefinition].
+
+        private val p = Promise[Done]()
+
+        override private[tasks] def run(recoverEnabled: Boolean)(implicit ec: ExecutionContext): Future[Done] = {
+          if (log.isDebugEnabled) {
+            log.debug("Performing task [{}] in CoordinatedShutdown phase [{}]", name, phaseName)
+          }
+          if (p.isCompleted) {
+            Future.successful(Done)
+          } else {
+            val f = try {
+              task.apply().recover {
+                case NonFatal(exc) if recoverEnabled =>
+                  log.warning("Task [{}] failed in phase [{}]: {}", name, phaseName, exc.getMessage)
+                  Done
+              }
+            } catch { // in case task.apply() throws
+              case NonFatal(exc) if recoverEnabled =>
+                log.warning(
+                  "Task [{}] in phase [{}] threw an exception before its future could be constructed: {}",
+                  name,
+                  phaseName,
+                  exc.getMessage)
+                Future.successful(Done)
+              case NonFatal(exc) =>
+                Future.failed(exc)
+            }
+            p.completeWith(f)
+            f
+          }
+        }
+
+        override def cancel(): Boolean = {
+          val cancelled = p.trySuccess(Done)
+          // now release all cancelled tasks in phase
+          registeredPhases
+            .merge(phaseName, StrictPhaseDefinition.empty, (previous, incoming) => previous.merge(incoming))
+          if (cancelled && log.isDebugEnabled) {
+            log.debug("Successfully cancelled CoordinatedShutdown task [{}] from phase [{}].", name, phaseName)
+          }
+          cancelled
+        }
+
+        override def isCancelled: Boolean = p.isCompleted
+      }
+    }
+
+    private case class StrictPhaseDefinition(tasks: Set[TaskDefinition]) extends PhaseDefinition {
+      // This is a case class so that the update methods on ConcurrentHashMap can correctly deal with equality
+      override val size: Int = tasks.size
+
+      override def run(recoverEnabled: Boolean)(implicit ec: ExecutionContext): Future[Done] = {
+        Future.sequence(tasks.map(_.run(recoverEnabled))).map(_ => Done)(ExecutionContexts.sameThreadExecutionContext)
+      }
+
+      // This method may be run multiple times during the compare-and-set loop of ConcurrentHashMap, so it must be side-effect-free
+      def merge(other: StrictPhaseDefinition): StrictPhaseDefinition = {
+        val nextTasks = (tasks ++ other.tasks).filterNot(_.isCancelled)
```

Comment:
Filtering out cancelled taken care of by the `merge` triggered on cancelling, why is it also needed here?

Reply:
I'm not sure I understand the question. This is the method that performs the merge triggered on cancelling; it seemed cleaner to drop them here rather than perform a second update in `cancel`.

Reply:
Ah, not sure what I was thinking there.